src/Http/Request/RequestExecutor.ts

```diff
@@ -12,147 +12,76 @@ import {IRavenObject} from "../../Database/IRavenObject";
 import {Lock} from "../../Lock/Lock";
 import {ILockDoneCallback} from "../../Lock/LockCallbacks";
 import {GetTopologyCommand} from "../../Database/Commands/GetTopologyCommand";
-import {RavenException, InvalidOperationException, BadRequestException, AuthorizationException} from "../../Database/DatabaseExceptions";
 import {StringUtil} from "../../Utility/StringUtil";
 import {DateUtil} from "../../Utility/DateUtil";
 import {IResponse, IErrorResponse} from "../Response/IResponse";
 import {StatusCodes, StatusCode} from "../Response/StatusCode";
+import {Observable} from "../../Utility/Observable";
+import {NodeSelector} from "./NodeSelector";
+import {RavenException, InvalidOperationException, BadRequestException, AuthorizationException, TopologyNodeDownException, AllTopologyNodesDownException} from "../../Database/DatabaseExceptions";
+
+export class RequestExecutor extends Observable {
+  public static readonly REQUEST_FAILED      = 'request:failed';
+  public static readonly TOPOLOGY_UPDATED    = 'topology:updated';
+  public static readonly NODE_STATUS_UPDATED = 'node:status:updated';
 
-export class RequestExecutor {
   protected conventions?: DocumentConventions;
   protected headers: IHeaders;
   private _lock: Lock;
   private _authenticator: ApiKeyAuthenticator;
-  private _topology: Topology;
   private _apiKey?: string = null;
   private _unauthorizedHandlerInitialized: boolean = false;
-  private _initUrl: string;
-  private _initDatabase: string;
+  private _firstTopologyUpdateCompleted: boolean = false;
+  private _disableTopologyUpdates: boolean;
+  private _nodeSelector: NodeSelector;
+  private _initialUrls: string[];
+  private _initialDatabase: string;
+
+  public get initialDatabase(): string {
+    return this._initialDatabase;
+  }
 
-  constructor(url: string, database: string, apiKey?: string, conventions?: DocumentConventions) {
-    const serverNode: ServerNode = new ServerNode(url, database, apiKey);
+  constructor(urls: string[], database: string, apiKey?: string, conventions?: DocumentConventions) {
+    super();
 
-    this._apiKey = apiKey;
-    this._initUrl = url;
-    this._initDatabase = database;
-    this.conventions = conventions;
-    this._lock = Lock.getInstance();
-    this._authenticator = new ApiKeyAuthenticator();
-    this._topology = new Topology(Number.MIN_SAFE_INTEGER, [serverNode]);
     this.headers = {
       "Accept": "application/json",
       "Has-Api-key": TypeUtil.isNone(apiKey) ? 'false' : 'true',
-      "Raven-Client-Version": "4.0.0.0"
+      "Raven-Client-Version": "4.0.0.0",
     };
 
-    setTimeout(() => this.updateFailingNodesStatuses(), 60 * 1000);
-    conventions.topologyUpdatesEnabled && setTimeout(() => this.getReplicationTopology(), 1 * 1000);
+    this._apiKey = apiKey;
+    this._initialUrls = urls;
+    this._initialDatabase = database;
+    this.conventions = conventions;
+    this._lock = Lock.getInstance();
+    this._disableTopologyUpdates = urls.length < 2;
```

Comment:
This will not work. Needs to be controlled by the convention only. Imagine a situation when someone will specify 1 url and later on will add a server to the cluster. We want the client to detect that and work properly.